[Accessibility] Inform screen reader users when mini cart updates

Manussakis opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #37597 
Closes #37598 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Open the shop page.
2. Turn on the screen reader (SR).
3. Go to a product card and click on the "Add to cart" button.
![image](https://github.com/woocommerce/woocommerce/assets/9420947/63acba7f-72a4-4a02-854b-9ae89956d642)
4. Suppose the Product's name is "Album", verify the SR announces "'Album' has been added to your cart"
5. Click on the same button again.
6. Verify the SR announces the same message again: "'Album' has been added to your cart".
7. Go to any other product and click on the "Add to cart" button".
8. Verify the SR announces the message with the name of the newly added product.
9. Go to the mini cart.
10. Remove a product.
![image](https://github.com/woocommerce/woocommerce/assets/9420947/072546e4-07da-4970-8724-aa0e26504546)
11. Assuming the removed product is the "Album" one, verify the SR announces "'Album' has been removed from your cart".
12. Remove another product and verify the message is updated with the name of the newly removed product.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Manussakis committed:
Create add_to_cart_success_message method in WC_Product_Simple class

Manussakis committed:
Add data-success_message attribute to remove item button in cart

Manussakis committed:
Add data-success_message to Add to cart button

Manussakis committed:
Make screen readers announce mini cart updates

Manussakis committed:
Add changelog file related to #37597 #37598

Manussakis committed:
Add translator comment to remove button data-success_message attribute

Manussakis committed:
Add docblock to woocommerce_product_add_to_cart_success_message filter

Manussakis committed:
Bump mini-cart template version

Manussakis committed:
Merge branch 'trunk' into feature/37597-37598-inform-screen-reader-users-when-mini-cart-updates

Manussakis commented:
@dkotter This PR is ready for code review 🙂

dkotter commented:
@ankitguptaindia @qasumitbagthariya This is ready for QA

ankitguptaindia commented:
Thanks @Manussakis for creating this PR. I found some scenarios where it is not working as expected.

**Case-1:** When there is no product added to the cart. 

When the cart is empty and the user adds a product to the cart, the screen reader announces nothing about the product being added. However, when the user removes the product from the cart, it announces both messages together: 'Product has been added to your cart' and 'Product has been removed from your cart'.




https://github.com/woocommerce/woocommerce/assets/6384299/07b4a8a1-583b-4a60-98b3-6bf949d850ab



**Case-2:** When the same product is already added to the cart. It announces the message, 'Product has been added to your cart' but also repeat same message when removing from cart. 


https://github.com/woocommerce/woocommerce/assets/6384299/163dd99d-fd6a-4593-a27d-a14a41f0deda




Please let me know if I am missing anything.

Manussakis committed:
Merge branch 'trunk' into feature/37597-37598-inform-screen-reader-users-when-mini-cart-updates

Manussakis committed:
Refresh mini cart live region by removing it from the DOM

Manussakis committed:
Add mini cart live region to DOM beforehand

Manussakis commented:
@ankitguptaindia Thanks for testing this PR.
I was not able to reproduce the issues you flagged. I have researched more about the `role="status"` attribute and figured out that sometimes screen readers don't announce the message when the live region element is added to the DOM at the same time as the message. 

Reference: [Why the role="status" attribute doesn't work correctly?](https://stackoverflow.com/questions/75463994/why-the-role-status-attribute-doesnt-work-correctly)

I updated the code to insert the live region element to the DOM beforehand and then the messages are appended to it as the user interacts with the mini cart.

@dkotter These are the two commits that I have added since your last review:
- [Refresh mini cart live region by removing it from the DOM](https://github.com/woocommerce/woocommerce/pull/48295/commits/d02186755b463d0973452b1a06f60ca93846b5b7)
- [Add mini cart live region to DOM beforehand](https://github.com/woocommerce/woocommerce/pull/48295/commits/29335a4218e01ffdc46d59c8ed7929f6f36779a8)

ankitguptaindia commented:
Update: Tested the same scenario with another browser's screen reader ([Screen Reader](https://chromewebstore.google.com/detail/screen-reader/kgejglhpjiefppelpmljglcjbhoiplfn))  This Screen reader announces correctly as expected. It looks like above reported issue related to macOS inbuilt voice reader. 



https://github.com/woocommerce/woocommerce/assets/6384299/a030305a-127f-4fe8-be22-b28068fb27b2

ankitguptaindia commented:
@vikrampm1 

Update: This PR works fine with  Screen Reader Chrome extension but with macOS's Voice-over, it still announces message about product addition at the time of product deletion from the cart. I am discussing this with @Manussakis in DM. 
Once we have any update, move this PR to the next step. Thank you! 

Quick Explanation: https://somup.com/cZ1YXlEYP3

Manussakis committed:
Merge branch 'woocommerce:trunk' into feature/37597-37598-inform-screen-reader-users-when-mini-cart-updates

Manussakis committed:
Remove aria-relevant attribute before adding new content to live region

Manussakis committed:
Merge branch 'woocommerce:trunk' into feature/37597-37598-inform-screen-reader-users-when-mini-cart-updates

Manussakis committed:
Add a delay before updating live region

Manussakis commented:
After syncing with @ankitguptaindia we realized that the issue was that the remove item action on the local environment was so fast that VoiceOver didn't have time to identify the updates in the live region.

@dkotter I have added a delay between messages along with this commit, please review it:
[Add a delay before updating live region](https://github.com/woocommerce/woocommerce/pull/48295/commits/f2113a1d6eca01c1bee0677b97c4622a7890d86c)

Manussakis commented:
@ankitguptaindia This is ready for QA 😉

ankitguptaindia commented:
Hi @Manussakis, I checked this PR after the latest fix. With my regular test environment and internet speed, I still see the message of adding and deleting products together. But if I reduce the speed (3GNetwork) by using Network Throttling in browser dev tool. It works great for me.


**With Slow 3G speed:**

https://github.com/woocommerce/woocommerce/assets/6384299/9c7ad373-1e1c-4ee0-9de1-d3c11e56d525

**With my regular network**

https://github.com/woocommerce/woocommerce/assets/6384299/2b8c29c9-e547-42ea-8d23-cf24d154969f



@Manussakis, could we proceed with this fix?

ankitguptaindia commented:
>After syncing with @ankitguptaindia we realized that the issue was that the remove item action on the local environment was so fast that VoiceOver didn't have time to identify the updates in the live region.



This PR functions as expected after the recent updates. I successfully tested it with a live URL (Local tunnel enabled Live URL), and the issue reported [here above in the thread](https://github.com/woocommerce/woocommerce/pull/48295#issuecomment-2195062216) is now resolved.


Quick Demo: 

https://github.com/woocommerce/woocommerce/assets/6384299/8765829c-caa7-42a6-a3bd-81a21fa425b6

Thanks for Great work here @Manussakis 🎉

vikrampm1 commented:
@coreymckrill this is ready for Woo review.

@Manussakis please remove [WIP] from the title.

jeffpaul commented:
@woocommerce/woo-fse note that this PR is ready for your review

danielwrobert commented:
Thanks @jeffpaul (and @coreymckrill)!

I believe this should be reviewed by @woocommerce/rubik, as they have taken ownership of the Mini-Cart.

jeffpaul commented:
@danielwrobert apologies as the woo-fse team label was on the related issue.  @woocommerce/rubik that means this PR is now up for your review, thanks!

danielwrobert commented:
> apologies as the woo-fse team label was on the related issue

@jeffpaul No need to apologize, FSE _used_ to own the mini cart and that shifted over to Rubic not too long ago. 🙂

wavvves committed:
Merge branch 'trunk' into feature/37597-37598-inform-screen-reader-users-when-mini-cart-updates

wavvves commented on plugins/woocommerce/templates/cart/mini-cart.php:
```diff
@@ -14,7 +14,7 @@
  * @see     https://woocommerce.com/document/template-structure/
  * @package WooCommerce\Templates
- * @version 7.9.0
+ * @version 9.1.0
```
```suggestion
 * @version 9.2.0
```

wavvves commented on plugins/woocommerce/includes/class-wc-product-simple.php:
```diff
@@ -74,4 +74,27 @@ public function add_to_cart_description() {
+
+		if ( ! empty( $text ) ) {
+			$text = sprintf( $text, $this->get_name() );
+		}
```
We can probably avoid the `empty()` check and make this more readable by encapsulating with the ternary condition:

```suggestion
        $text = '';
        if ( $this->is_purchasable() && $this->is_in_stock() ) {
            /* translators: %s: Product title */
            $text = __( '&ldquo;%s&rdquo; has been added to your cart', 'woocommerce' );
            $text = sprintf( $text, $this->get_name() );
        }
```

Manussakis replied:
@wavvves Please review it: [Refactor logic in the add_to_cart_success_message function](https://github.com/woocommerce/woocommerce/pull/48295/commits/728af717e0dc75479b50e9d88b2756f9d3647171)

wavvves commented on plugins/woocommerce/includes/class-wc-product-simple.php:
```diff
@@ -74,4 +74,27 @@ public function add_to_cart_description() {
+		/**
+		 * Filter product add to cart success message.
+		 *
+		 * @since 8.9.2
```
This is a newly introduced filter, correct?

```suggestion
		 * @since 9.2.0
```

wavvves commented on plugins/woocommerce/includes/class-wc-product-simple.php:
```diff
@@ -74,4 +74,27 @@ public function add_to_cart_description() {
+		 * Filter product add to cart success message.
+		 *
+		 * @since 8.9.2
+		 * @param string $text.
```
We might want to add a description for the `$text` variable here.

Manussakis replied:
@wavvves Please review it: [Add description to the $text param on woocommerce_product_add_to_cart…](https://github.com/woocommerce/woocommerce/pull/48295/commits/ea52a84d8604587924ad0c08e9c330f93a552818)

wavvves reviewed (changes requested):
Hi @Manussakis, thank you for your PR. I left some comments about versioning and a nitpick refactor to avoid an unnecessary if clause and improve readability 👍🏼

Manussakis committed:
Bump mini-cart template version from 9.1.0 to 9.2.0

Co-authored-by: Paulo Arromba <17236129+wavvves@users.noreply.github.com>

Manussakis committed:
Merge branch 'trunk' into feature/37597-37598-inform-screen-reader-users-when-mini-cart-updates

Manussakis committed:
Bump version of woocommerce_product_add_to_cart_success_message filter

Co-authored-by: Paulo Arromba <17236129+wavvves@users.noreply.github.com>

Manussakis committed:
Add description to the $text param on woocommerce_product_add_to_cart_success_message filter

Manussakis committed:
Refactor logic in the add_to_cart_success_message function

Manussakis commented:
@wavvves Thanks for reviewing the PR. I have addressed your feedback, please re-review it.

wavvves committed:
Merge branch 'trunk' into feature/37597-37598-inform-screen-reader-users-when-mini-cart-updates

wavvves reviewed (approved):
Working as described, thank you @Manussakis

vikrampm1 commented:
@wavvves this has completed all the steps from our end and is ready to be merged by Woo.

ralucaStan commented on plugins/woocommerce/client/legacy/js/frontend/add-to-cart.js:
```diff
@@ -207,6 +218,55 @@ jQuery( function( $ ) {
+	 * Update cart live region message after add/remove cart events.
+	 */
+	AddToCartHandler.prototype.alertCartUpdated = function( e, fragments, cart_hash, $button ) {
+		var message = $button.data( 'success_message' );
```
This part breaks the Add to cart functionality for extensions that are calling this function with the 4th parameter. We received reports of this after the release of WooCommerce 9.3.

Manussakis replied:
Hey @ralucaStan 

Would you mind sending me the link to the issue so I can better understand the regression?

Thanks!

samueljseay replied:
@Manussakis we have addressed it in a follow up PR: https://github.com/woocommerce/woocommerce/pull/51449

It turns out 3PD extensions will trigger the event manually sometimes and won't pass the button element in those cases. I put a fuller description on the PR.

Manussakis replied:
Thank you, @samueljseay !
